python_ta/checkers/redundant_assignment_checker.py

```diff
@@ -0,0 +1,126 @@
+    name = 'redundant_assignment'
+    # use dashes for connecting words in message symbol
+    msgs = {'E9959': ('This assignment statement is redundant;' \
+                      ' You can remove it from the program.',
+                      'redundant-assignment',
+                      'This assignment statement is redundant;' \
+                      ' You can remove it from the program.')
+           }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    def __init__(self, linter=None):
+        super().__init__(linter=linter)
+        self._redundant_assignment: Set[astroid.Assign] = set()
+
+    @check_messages('redundant-assignment')
+    def visit_assign(self, node: astroid.Assign):
+        if node in self._redundant_assignment:
+            self.add_message('redundant-assignment', node=node)
+
+    def visit_module(self, node: astroid.Module):
+        self._analyze(node)
+
+    def visit_functiondef(self, node: astroid.FunctionDef):
+        self._analyze(node)
+
+    def _analyze(self, node: Union[astroid.Module, astroid.FunctionDef]) -> None:
+        """Runs the backward data flow algorithm on a `Module` or `Function` CFG, which in turn
+        appends `Assign` nodes to `self.redundant_assignment` if it is redundant.
+
+        Data flow algorithms retrieved from:
+        https://www.seas.harvard.edu/courses/cs252/2011sp/slides/Lec02-Dataflow.pdf#page=31
+        """
+        # Stores all the variable names that will be re-defined before any usage at a
+        # particular program point.
+        out_facts = {}
+        cfg = ControlFlowGraph()
+        cfg.start = node.cfg_block
+        worklist = list(cfg.get_blocks_postorder())
+        worklist.reverse()
+
+        all_assigns = self._get_assigns(node)
+        for block in worklist:
+            out_facts[block] = all_assigns.copy()
+
+        while len(worklist) != 0:
+            b = worklist.pop()
+            outs = [out_facts[p.target] for p in b.successors if p.target in out_facts]
+            if outs == []:
+                in_facts = set()
+            else:
+                in_facts = set.intersection(*outs)
+            temp = self._transfer(b, in_facts)
+            if temp != out_facts[b]:
+                out_facts[b] = temp
+                worklist.extend([pred.source for pred in b.predecessors])
+
+    def _transfer(self, block: CFGBlock, out_facts: Set[str]) -> Set[str]:
+        gen = out_facts.copy()
+        kill = set()
+        for statement in reversed(block.statements):
+            if isinstance(statement, astroid.FunctionDef):
+                # `nodes_of_class` below does block looking for required nodes
+                # in function definitions, hence this case.
+                continue
+            for node in statement.nodes_of_class((astroid.AssignName, astroid.DelName, astroid.Name,
+                                                  astroid.Nonlocal, astroid.Global),
```

Comment:
Okay it's fine to include these two node classes nodes here, but they need a separate case to remove names from `kill`. `Global` and `NonLocal` nodes don't have a `name` attribute, they have a `names` attribute which is a `List[str]`.